### PR TITLE
Initial commit for the React port of ui-predicate-vue

### PR DESCRIPTION
**Pull request by mardii**
### Prerequisites
- [x] I have read the [Contributing guidelines](https://github.com/fgribreau/ui-predicate/blob/master/.github/CONTRIBUTING.md)
- [x] I have read the [Code of conduct](https://github.com/fgribreau/ui-predicate/blob/master/.github/CODE_OF_CONDUCT.md) and I agree with it

### Description
Initial commit for the React port of `ui-predicate-vue`

- all of the Vue code - except for testing - is ported to React (perhaps somewhat naively - but it's working)
- storybook is up and running

Tried to use as much of the same structure and logic as in the `ui-predicate-vue` package. Perhaps there were some things that could've done smarter or more 'React-like'. But keeping more close to the Vue implementation was faster to do, plus it'll help with future development of both platforms.

**Issue :** [Reference to the issue]
Testing is not set up yet. There are some hints of storybook snapshots, but haven't tested.
Wasn't able to run storybook correctly for the vue package, so used the [online version](https://ui-predicate.fgribreau.com/ui-predicate-vue/0.6.2/) as a reference.

### Example usage
To see the react code in action:
```bash
npm run storybook
```

### CODE REVIEW

File: packages/ui-predicate-react/.storybook/preview-head.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+So, these won’t be loaded into the main Storybook UI. -->
+<style>
+  /* for demo purpose */
+  .ui-predicate.compound {
+    margin-left: 10px;
+  }
+</style>
+
+<style>
+  .custom-css .ui-predicate__main,
```
Comment (FGRibreau) [inline packages/ui-predicate-react/.storybook/preview-head.html]: should we move this CSS outside of both react & vuejs package?

File: packages/ui-predicate-react/src/ui-predicate-options.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+class UIPredicateOptions extends Component {
+  render() {
+    const {
+      predicate,
+      remove,
+      add,
+      getUIComponent,
+      getAddCompoundMode,
+    } = this.props;
+    const RemoveComponent = getUIComponent(UITypes.PREDICATE_REMOVE);
```
Comment (FGRibreau) [inline packages/ui-predicate-react/src/ui-predicate-options.js]: AddPredicate & RemovePredicate might be a better name?

File: packages/ui-predicate-react/src/ui-predicate.css [hunk 3]
```diff
@@ -0,0 +1,12 @@
+.ui-predicate__main {
```
Comment (FGRibreau) [inline packages/ui-predicate-react/src/ui-predicate.css]: we might want to share this CSS between react & vuejs (as well as other framework) right?

File: packages/ui-predicate-react/src/ui-predicate.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    const { isCoreReady, root } = this.state;
+    const { columns, className } = this.props;
+    return (
+      <div className={`ui-predicate__main ${className}`}>
+        {!isCoreReady && Boolean(this.props.loader) && this.props.loader}
+        {Boolean(isCoreReady) && (
+          <UIPredicateCompound
+            predicate={root}
+            columns={columns}
+            getAddCompoundMode={this.getAddCompoundMode}
```
Comment (FGRibreau) [inline packages/ui-predicate-react/src/ui-predicate.js]: UIPredicateCompound should only have `predicate` & `columns` props. 
Why does it have add, remove, setPredicateLogicalType_id etc.. here?

File: packages/ui-predicate-react/src/UIPredicateCoreReact.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+const defaults = {
+  options: {
+    /**
+     * UIPredicate React Adapter own default argument component
+     * @param {Object} [columns=PredicateCore.defaults.columns] columns
+     * @param {Object} [options=PredicateCore.defaults.options] options
+     * @param {Object} [ui=PredicateCore.defaults.ui] ui
+     * @return {Vue.component} the default Vue Component to use as argument specifier
+     * @see core.defaults.getDefaultArgumentComponent
+     * @memberof vue.defaults
```
Comment (davidlandais) [inline packages/ui-predicate-react/src/UIPredicateCoreReact.js]: Does it should belongs to react not vue ?

File: packages/ui-predicate-react/src/default-components/logical-types.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import React, { Component } from 'react';
+
+class LogicalTypes extends Component {
```
Comment (davidlandais) [inline packages/ui-predicate-react/src/default-components/logical-types.js]: You are not managing state inside this component, only props. You could use functionnal component

File: packages/ui-predicate-react/src/UIPredicateCoreReact.js (hunk 5 above)
Comment (mardii) [reply packages/ui-predicate-react/src/UIPredicateCoreReact.js]: Ooh yeah. That's was an incorrect copy-paste. I'll fix it.

File: packages/ui-predicate-react/src/default-components/logical-types.js (hunk 6 above)
Comment (mardii) [reply packages/ui-predicate-react/src/default-components/logical-types.js]: You are right. I will change it.

File: packages/ui-predicate-react/.storybook/preview-head.html (hunk 1 above)
Comment (mardii) [reply packages/ui-predicate-react/.storybook/preview-head.html]: You decide.
We could leave it as is for now. And then in the next PR see if we can move it to a central place. I'm still working on getting the CSS to nicely build in.

File: packages/ui-predicate-react/src/ui-predicate.js (hunk 4 above)
Comment (mardii) [reply packages/ui-predicate-react/src/ui-predicate.js]: I will have to look into it again to be sure. In the Vue version, the properties are injected. In React we need to pass them through the hierarchy manually.

**Comment by mardii** [comment]
Hey Guys. Thanks for the feedback! 

First to answer your question:
I wanted to port this package because I want to use it for a project indeed. 

It will take some days to still fix some remaining things (mostly related to the building of the package) and processing your feedback.

In retrospect, I should've waited before opening the PR before I tackled some of the building stuff I'm working on now. On the other hand, I felt it was good to ask for feedback early on.

I will send a new commit when I'm able to build all the parts, and getting it to run in my own project. Hopefully end of the week.

File: packages/ui-predicate-react/src/ui-predicate.js (hunk 4 above)
Comment (mardii) [reply packages/ui-predicate-react/src/ui-predicate.js]: So yeah... These properties are passed along into the components manually. 
I could rewrite the code a bit to make it less verbose (picking the function from `this.state.ctrl` and then destructure them into the component). Or... I could look into how `context` works in React. I read that it works in similar ways as `inject` in *Vue*. But haven't worked with `context` before (still new-ish to React).
What do you think?

File: packages/ui-predicate-react/.storybook/preview-head.html (hunk 1 above)
Comment (mardii) [reply packages/ui-predicate-react/.storybook/preview-head.html]: Actually... thinking about this. This is really bound to the storybook, right?
With every view implementation has its own repo, it also has its own storybook. So sharing the CSS between them kind of couples the individual packages in a way. On the other hand... this is a bit of repeating yourself (no so DRY - can I say WET? )

So what do you think?

File: packages/ui-predicate-react/src/ui-predicate.css (hunk 3 above)
Comment (mardii) [reply packages/ui-predicate-react/src/ui-predicate.css]: This CSS I might want to share yes.
So far I haven't had any luck including the CSS correctly when using the package. I need to look into it. 

Where do you think we should place the css in this case?
And do you think we should do this in this PR, or open up a second one for it?

Comment (FGRibreau) [reply packages/ui-predicate-react/src/ui-predicate.css]: Indeed, we can do this later don't worry :)

**Comment by FGRibreau** [comment]
> I wanted to port this package because I want to use it for a project indeed.

@mardii  Awesome :)


> I will send a new commit when I'm able to build all the parts, and getting it to run in my own project. Hopefully end of the week.

@mardii Perfect!

File: packages/ui-predicate-react/src/ui-predicate.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
+import React, { Component } from 'react';
+import fs from 'fs';
+import UIPredicateCoreReact from './UIPredicateCoreReact';
+import { InitializationFailed } from './errors';
+import UIPredicateCompound from './ui-predicate-compound';
+import './ui-predicate-react.css';
```
Comment (davidlandais) [inline packages/ui-predicate-react/src/ui-predicate.js]: Mardii, i don't know well ReactJS in a browser context so, why did you import style here and then read the file content and appending it to a new node ?

**Comment by FGRibreau** [comment]
superseded by https://github.com/FGRibreau/ui-predicate/pull/132

**Pull request closed without merging**